internal/runtime/storeforward.go

```diff
@@ -0,0 +1,243 @@
+	config *common.ConfigurationStruct,
+	edgeXClients common.EdgeXClients) {
+
+	items, err := sf.storeClient.RetrieveFromStore(serviceKey)
+	if err != nil {
+		edgeXClients.LoggingClient.Error("Unable to load store and forward items from DB", "error", err)
+		return
+	}
+
+	edgeXClients.LoggingClient.Debug(fmt.Sprintf(" %d stored data items found for retrying", len(items)))
+
+	if len(items) > 0 {
+		itemsToRemove, itemsToUpdate := sf.processRetryItems(items, config, edgeXClients)
+
+		edgeXClients.LoggingClient.Debug(
+			fmt.Sprintf(" %d stored data items will be removed post retry", len(itemsToRemove)))
+		edgeXClients.LoggingClient.Debug(
+			fmt.Sprintf(" %d stored data items will be update post retry", len(itemsToUpdate)))
+
+		for _, item := range itemsToRemove {
+			if err := sf.storeClient.RemoveFromStore(item); err != nil {
+				edgeXClients.LoggingClient.Error(
+					"Unable to remove stored data item from DB",
+					"error", err,
+					"objectID", item.ID,
+					clients.CorrelationHeader, item.CorrelationID)
+			}
+		}
+
+		for _, item := range itemsToUpdate {
+			if err := sf.storeClient.Update(item); err != nil {
+				edgeXClients.LoggingClient.Error("Unable to update stored data item in DB",
+					"error", err,
+					"objectID", item.ID,
+					clients.CorrelationHeader, item.CorrelationID)
+			}
+		}
+	}
+}
+
+func (sf *storeForwardInfo) processRetryItems(items []contracts.StoredObject,
+	config *common.ConfigurationStruct,
+	edgeXClients common.EdgeXClients) ([]contracts.StoredObject, []contracts.StoredObject) {
+
+	var itemsToRemove []contracts.StoredObject
+	var itemsToUpdate []contracts.StoredObject
+
+	for _, item := range items {
+		if item.Version == sf.calculatePipelineHash() {
+			if !sf.retryExportFunction(item, config, edgeXClients) {
+				item.RetryCount++
+				if config.Writable.StoreAndForward.MaxRetryCount == 0 ||
+					item.RetryCount < config.Writable.StoreAndForward.MaxRetryCount {
+					edgeXClients.LoggingClient.Trace("Export retry failed. Incrementing retry count",
+						"retries",
+						item.RetryCount,
+						clients.CorrelationHeader,
+						item.CorrelationID)
+					itemsToUpdate = append(itemsToUpdate, item)
+					continue
+				}
+
+				edgeXClients.LoggingClient.Trace(
+					"Max retries exceeded. Removing item from DB", "retries",
+					item.RetryCount,
+					clients.CorrelationHeader,
+					item.CorrelationID)
+				// Note that item will be removed for DB below.
+			} else {
+				edgeXClients.LoggingClient.Trace(
+					"Export retry successful. Removing item from DB",
+					clients.CorrelationHeader,
+					item.CorrelationID)
+			}
+		} else {
+			edgeXClients.LoggingClient.Error(
+				"Stored data item's Function Pipeline Version doesn't match current Function Pipeline Version. Removing item from DB",
+				clients.CorrelationHeader,
+				item.CorrelationID)
+		}
+
+		// Will remove from store if version no longer matches current Pipeline or max retries exceeded
```

Comment:
I might call out the continue statement above in the comments -- looks like everything always gets removed with this here.

Reply:
Good point. I have revamp that comment for readability.